llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1207,6 +1207,18 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
     }
   }
 
+  // (X * C0) / (X * C1) --> C0 / C1
+  Constant *C0, *C1;
+  if (match(Op0, m_c_Mul(m_Value(X), m_Constant(C0))) &&
```

Comment:
I think we can generalize this pattern to handle `sdiv (X *nsw Y), (X *nsw Z) -> sdiv Y, Z` iff `sdiv Y, Z` doesn't overflow.
Alive2: https://alive2.llvm.org/ce/z/5TdmfG

Reply:
What do I need to do

Reply:
> What do I need to do

You should check `c1 != -1 || c1 != sign mask`.